extensions/amp-skimlinks/0.1/link-rewriter/link-rewriter.js

```diff
@@ -0,0 +1,239 @@
+   * @param {string} id
+   * @param {function(Array<HTMLElement>):TwoStepsResponse} resolveUnknownLinks
+   * @param {?{linkSelector: string}=} options
+   */
+  constructor(ampPageDocument, id, resolveUnknownLinks, options) {
+    /** @public {!./event-messenger.EventMessenger} */
+    this.events = new EventMessenger();
+
+    /** @public {string} */
+    this.id = id;
+
+    /** @private {Document} */
+    this.ampPageDocument_ = ampPageDocument;
+
+
+    /** @private {function(Array<HTMLElement>):TwoStepsResponse} */
+    this.resolveUnknownLinks_ = resolveUnknownLinks;
+
+    /** @private {string} */
+    this.linkSelector_ = options.linkSelector || 'a';
+
+    /** @private {number} */
+    this.restoreDelay_ = 300; //ms
+
+    /** @private {!./link-replacement-cache.LinkReplacementCache} */
+    this.anchorReplacementCache_ = new LinkReplacementCache();
+  }
+
+  /**
+   * @public
+   * Get the replacement url associated with the anchor.
+   * @param {?HTMLElement} anchor
+   * @return {?string}
+   */
+  getReplacementUrl(anchor) {
+    if (!this.isWatchingLink(anchor)) {
+      return null;
+    }
+
+    return this.anchorReplacementCache_.getReplacementUrlForAnchor(anchor);
+  }
+
+  /**
+   * @public
+   * Get the anchor to replacement url cache.
+   * Useful to extract information for tracking purposes.
+   * @return {!AnchorReplacementList}
+   */
+  getAnchorReplacementList() {
+    return this.anchorReplacementCache_.getAnchorReplacementList();
+  }
+
+  /**
+   * @public
+   * Returns True if the LinkRewriter instance is supposed to handle
+   * this particular anchor.
+   * By default LinkRewriter handles all the links on the page but
+   * inclusion/exclusion rules can be created by using the "linkSelector"
+   * option. When using this option all the links not matching the css selector
+   * will be ignored and isWatchingLink(anchor) will return false.
+   * @param {?HTMLElement} anchor
+   * @return {boolean}
+   */
+  isWatchingLink(anchor) {
+    return this.anchorReplacementCache_.isInCache(anchor);
+  }
+
+  /**
+   * @public
+   * This function is called when the user clicks on a link.
+   * It swaps temporarly the href of an anchor by its associated
+   * replacement url but only for the time needed by the browser
+   * to handle the click on the anchor and navigate to the new url.
+   * After 300ms, if the page is still open (target="_blank" scenario),
+   * the link is restored to its initial value.
+   * @param {?HTMLElement} anchor
+   * @return {boolean} - 'true' if the linkRewriter has changed the url
+   *  'false' otherwise.
+   */
+  rewriteAnchorUrl(anchor) {
+    const newUrl = this.getReplacementUrl(anchor);
+    if (!newUrl || newUrl === anchor.href) {
+      return false;
+    }
+    // Save so we can restore it.
+    anchor.setAttribute(ORIGINAL_URL_ATTRIBUTE, anchor.href);
+    anchor.href = newUrl;
+    // Restore link to original after X ms.
+    setTimeout(() => {
```

Comment:
hm..I am wondering if it's safer and simpler to just use window.open to open these monetized links rather than replacing them and then resetting it back to its original value.

Reply:
We use this technique of temporarily replacing the link in Skimlinks normal javascript and it works well. This is the cleanest/most natural solution since we let the browser handle the click without risking to interfere with the browser default behaviour.

Reply:
Right, so lets not deviate from that behavior then.  I guess we would also be misleading the user if they hover over the link and when clicked, was directed to the monetized link rather than what the original link, as it was not mutated..so replacing the link is the right behavior. Thanks.